# E-commerce Customer Profiling and Data Consolidation Project

# Problem Statement

An e-commerce startup, after a successful first month, is seeking to better understand its customer base and purchasing patterns. They aim to answer fundamental questions like "Who are our customers?" "What do they buy?" and "What drives their purchasing behavior?" However, they face challenges in accurately identifying and counting customers due to disparate data sources, including:

1. A customer database recording online account sign-ups,
2. A CRM system tracking phone and non-online customer interactions, and
3. Raw transaction data that includes guest purchases with no formal customer record.

These sources of data may have overlapping entries or duplicate records, often due to customers engaging with multiple systems (e.g., making a purchase as a guest, then creating an account later). Furthermore, customer information may vary across entries due to potential discrepancies like typos or alternative identifiers. The startup seeks the assistance of an analyst to address these complexities and provide an accurate count and profile of their customers.


## Project Goals

1. **Data Consolidation:**  
   Integrate customer information across all available sources to create a unified, comprehensive dataset representing the complete customer base.

2. **Data Deduplication:**  
   Identify and resolve duplicate records, maintaining data lineage to track original sources and identifiers for each customer where possible.

3. **Customer Identification:**  
   Develop a data model that uniquely identifies each customer, regardless of which data source(s) they appear in, and can be easily queried for counts and profiles.

4. **Data Validation:**  
   Ensure data integrity by confirming assumptions, identifying inconsistencies, and validating data completeness to maximize reliability.

5. **Solution Documentation:**  
   Create a flexible and traceable solution, with a well-defined schema that supports future analysis of customer demographics and purchasing behavior.

This project will ultimately enable the startup to answer critical questions about their customer demographics and purchasing trends, creating a foundation for data-driven decision-making.

To actually perform the data modeling, we should take the following course of action:

1. First we explore all 3 datasets. We should see if the values in all the columns make sense.

2. Then we can trim down our data before we merge our different datasets, so that we have unique customers

3. Then we merge these trimmed datasets and remove duplicates carefully.

4. When performing deduplication, we need to go one step beyond and identify more than just exact match duplicates. We can use Fuzzy Matching here.

5. Finally, we clean and present our data model.

## Exploring our Purchases Dataset

In [1]:
import pandas as pd
import numpy as np

sales = pd.read_csv("../data/purchases.csv")
print(sales.shape)

(71519, 11)


In [2]:
sales.head()


event_time  product_id          category_id  \
0  2022-10-01 02:26:08+00:00    32701106  2055156924466332447   
1  2022-10-01 02:28:32+00:00     9400066  2053013566067311601   
2  2022-10-01 02:31:01+00:00     1004238  2053013555631882655   
3  2022-10-01 02:33:31+00:00    11300059  2053013555531219353   
4  2022-10-01 02:40:18+00:00    17300751  2053013553853497655   

            category_code    brand    price  \
0                     NaN  shimano    95.21   
1                     NaN   jaguar   164.20   
2  electronics.smartphone    apple  1206.40   
3   electronics.telephone    texet    17.48   
4                     NaN  versace    77.22   

                             session_id  customer_id guest_first_name  \
0  64c68405-7002-4ce0-9604-a4c2e1f7384b          NaN          MICHAEL   
1  3b7d6741-3c82-4c75-8015-6f54b52612e0       7466.0              NaN   
2  38c6d3f7-6c32-4fed-bca6-ef98e1746386          NaN             COLE   
3  3398c966-7846-4186-89be-323daad735b9          NaN         MOHAMMED   
4  11e3a573-01b9-4794-b513-e7d8a4fcac83      31266.0              NaN   

  guest_surname guest_postcode  
0         MASON        RG497ZQ  
1           NaN            NaN  
2     WILKINSON         SW75TQ  
3      RICHARDS        RG150RE  
4           NaN            NaN

In [3]:
sales.isnull().sum()

event_time              0
product_id              0
category_id             0
category_code       16739
brand                5707
price                   0
session_id              0
customer_id         18448
guest_first_name    53071
guest_surname       53071
guest_postcode      53071
dtype: int64

Looks like the 18448 missing customer IDs (which are guest checkouts) and the 53071 missing guest values (which are registered customers) make up all of our data. We should verify there's no overlap, e.g. a row with both customer ID and guest details missing.

Let's create a column to track guest checkouts:

In [4]:
sales["is_guest"] = sales["customer_id"].isnull()

Let's check cases where a guest checkout also had a customer id filled

In [5]:
sales[sales["is_guest"] & sales["customer_id"].notnull()]

Empty DataFrame
Columns: [event_time, product_id, category_id, category_code, brand, price, session_id, customer_id, guest_first_name, guest_surname, guest_postcode, is_guest]
Index: []

no such case. Let's also check cases where we neither have a guest checkout nor a customer id filled.

In [6]:
sales[(sales["is_guest"] == False) & sales["customer_id"].isnull()]

Empty DataFrame
Columns: [event_time, product_id, category_id, category_code, brand, price, session_id, customer_id, guest_first_name, guest_surname, guest_postcode, is_guest]
Index: []

This tells us that either all rows are a guest checkout or a purchase made by a registered customer. Let's now check what percentage of records are guest checkouts.

In [7]:
sales['is_guest'].value_counts(normalize=True)

False    0.742055
True     0.257945
Name: is_guest, dtype: float64

Now we know around 25% of our records are guest checkouts. We need to keep in mind that each row in our dataset represents a purchased item and not a customer record. Hence, this won't be the actual proportion of guest checkouts. Now, to calculate the actual proportion:

In [8]:
guest_columns = ["guest_first_name", "guest_surname", "guest_postcode"]
unique_guests = sales[guest_columns].drop_duplicates()
print(len(unique_guests))

unique_customers = sales["customer_id"].unique()
cust_total = len(unique_customers) + len(unique_guests) - 1 # Subtracting 1 because Null is also counted once
print(len(unique_customers))
print(cust_total)

8301
24962
33262


## Exporting Customer Data from Purchases

We need to create a schema where we can store our customer records. If we look at our datasets, we can see that customer and CRM datasets have names, postcodes and age (which is not present in our guest customers, but we should not discard this column just because its not present in guests). We can also keep track of where our records came from by adding an indicator column for each data source. We could also just have one column called sources with values like 'purchases', 'CRM' or 'customer database' but it would not help us if we have duplicate customers in different sources.

Our data model schema will contain these columns:
customer_id, first_name, surname, postcode, age, is_guest, in_purchase_data, in_crm_data, in_customer_data

To extract customer data from purchases, we should first extract the guests and non-guests separately and then combine both.

In [9]:
guest_columns = ['guest_first_name', 'guest_surname', 'guest_postcode', 'is_guest']
guests = sales.loc[sales['is_guest'], guest_columns]
guests = guests.drop_duplicates()
guests.head()

guest_first_name guest_surname guest_postcode  is_guest
0           MICHAEL         MASON        RG497ZQ      True
2              COLE     WILKINSON         SW75TQ      True
3          MOHAMMED      RICHARDS        RG150RE      True
7              KIAN         MILLS        SW332TF      True
13             RUBY          OWEN        PO377YS      True

In [10]:
non_guests = (pd.DataFrame(sales.loc[sales["customer_id"].notnull(), "customer_id"]
                          .unique()
                          .astype(int),
                          columns = ['customer_id']
                          ))
non_guests.head()

customer_id
0         7466
1        31266
2    534142828
3         1035
4         6985

In [11]:
sales_customers = pd.concat([non_guests, guests], axis=0, ignore_index=True)

In [12]:
new_col_names = ["customer_id", "first_name", "surname","postcode", "is_guest"]
sales_customers = sales_customers.set_axis(new_col_names, axis=1)

In [13]:
sales_customers["is_guest"] = sales_customers["is_guest"].fillna(False)

In [14]:
sales_customers["in_purchase_data"] = True

In [15]:
for col in ["first_name", "surname"]:
    sales_customers[col] = sales_customers[col].str.lower().str.strip()


sales_customers["postcode"] = sales_customers["postcode"].str.strip()

We have successfully extracted customers data from purchases.

## Exploring CRM Data

In [16]:
crm = pd.read_csv("../data/crm_export.csv")
print(crm.shape)
crm.head()

(7825, 5)


customer_id first_name  surname postcode  age
0        29223      Holly   Rogers  LS475RT   12
1        27826     Daniel     Owen   M902XX    5
2         7432    Eleanor  Russell  HR904ZA   34
3         2569      Paige  Roberts  DE732EP   61
4         9195    Matilda    Young  LS670FU   78

In [17]:
crm.isnull().sum()


customer_id    0
first_name     0
surname        0
postcode       0
age            0
dtype: int64

In [18]:
crm.isnull().sum()

customer_id    0
first_name     0
surname        0
postcode       0
age            0
dtype: int64

Sanitizing string columns

In [19]:
for col in ["first_name", "surname"]:
    crm[col] = crm[col].str.lower().str.strip()

crm["postcode"] = crm["postcode"].str.strip()

In [20]:
# Verifying that customer ID is unique
crm.groupby("customer_id").size().loc[lambda x: x > 1]

Series([], dtype: int64)

In [21]:
# Checking for duplicate information
print(len(crm))
print(len(crm.drop(columns="customer_id").drop_duplicates()))


7825
7419


In [22]:
print("{} rows".format(len(crm)))
print("{} unique combinations of customer information".format(len(crm.drop(columns="customer_id").drop_duplicates())))

7825 rows
7419 unique combinations of customer information


Now let's join CRM data to purchases where purchases have a customer ID (not guests)

In [23]:
print(len(sales_customers))
sales_customers.isnull().sum()

33261


customer_id          8300
first_name          24961
surname             24961
postcode            24961
is_guest                0
in_purchase_data        0
dtype: int64

In [24]:
sales_and_crm_customers = sales_customers.merge(crm, on="customer_id", how="left", suffixes=("_sales", "_crm"))
print(len(sales_and_crm_customers))
sales_and_crm_customers.isnull().sum()

33261


customer_id          8300
first_name_sales    24961
surname_sales       24961
postcode_sales      24961
is_guest                0
in_purchase_data        0
first_name_crm      26147
surname_crm         26147
postcode_crm        26147
age                 26147
dtype: int64

In [25]:
# now we have duplicate customer detail columns, so merge them into the old ones
merged_customers_filter = (
    (sales_and_crm_customers["customer_id"].notnull()) # only for actual customers
    # only if they have at least first or surname filled in
    # meaning we've found a match in our CRM data
    & ((sales_and_crm_customers["first_name_crm"].notnull())
       | (sales_and_crm_customers["surname_crm"].notnull()))
)

sales_and_crm_customers.loc[merged_customers_filter, "in_crm_data"] = True
sales_and_crm_customers.loc[~merged_customers_filter, "in_crm_data"] = False

sales_and_crm_customers["in_crm_data"].value_counts()

False    26147
True      7114
Name: in_crm_data, dtype: int64

Now merge data into single versions of customer information

In [26]:
sales_and_crm_customers.loc[merged_customers_filter, ["first_name_sales", "surname_sales", "postcode_sales"]] = (
    sales_and_crm_customers.loc[merged_customers_filter, ["first_name_crm", "surname_crm", "postcode_crm"]]
    .values
)

# drop duplicate columns and rename
sales_and_crm_customers = (
    sales_and_crm_customers
    .drop(columns=["first_name_crm", "surname_crm", "postcode_crm"])
    .rename(columns={
        "first_name_sales": "first_name",
        "surname_sales": "surname",
        "postcode_sales": "postcode"
    })
)

print(sales_and_crm_customers.isnull().sum())
sales_and_crm_customers.head()

customer_id          8300
first_name          17847
surname             17847
postcode            17847
is_guest                0
in_purchase_data        0
age                 26147
in_crm_data             0
dtype: int64


customer_id first_name surname postcode  is_guest  in_purchase_data   age  \
0       7466.0        NaN     NaN      NaN     False              True   NaN   
1      31266.0     harley  palmer  HR250EJ     False              True  33.0   
2  534142828.0        NaN     NaN      NaN     False              True   NaN   
3       1035.0        NaN     NaN      NaN     False              True   NaN   
4       6985.0        NaN     NaN      NaN     False              True   NaN   

  in_crm_data  
0       False  
1        True  
2       False  
3       False  
4       False

Now, we will add CRM customers not in our purchase history

In [27]:
crm_ids_to_add = list(set(crm["customer_id"].unique()) - set(sales_and_crm_customers["customer_id"].unique()))
print(len(crm_ids_to_add))

711


In [28]:
print(len(sales_and_crm_customers))

sales_and_crm_customers = (
    pd.concat([sales_and_crm_customers, crm[crm["customer_id"].isin(crm_ids_to_add)]],
              axis=0,
             ignore_index=True)
)

print(len(sales_and_crm_customers))

33261
33972


In [29]:
sales_and_crm_customers.isnull().sum()

customer_id          8300
first_name          17847
surname             17847
postcode            17847
is_guest              711
in_purchase_data      711
age                 26147
in_crm_data           711
dtype: int64

Updating missing record sources accordingly

In [30]:
sales_and_crm_customers["is_guest"] = sales_and_crm_customers["is_guest"].fillna(False)
sales_and_crm_customers["in_purchase_data"] = sales_and_crm_customers["in_purchase_data"].fillna(False)
sales_and_crm_customers["in_crm_data"] = sales_and_crm_customers["in_crm_data"].fillna(True)

sales_and_crm_customers.isnull().sum()

customer_id          8300
first_name          17847
surname             17847
postcode            17847
is_guest                0
in_purchase_data        0
age                 26147
in_crm_data             0
dtype: int64

In [31]:
sales_and_crm_customers["in_crm_data"].value_counts()

False    26147
True      7825
Name: in_crm_data, dtype: int64

In [32]:
print(len(sales_and_crm_customers))
sales_and_crm_customers.head()

33972


customer_id first_name surname postcode  is_guest  in_purchase_data   age  \
0       7466.0        NaN     NaN      NaN     False              True   NaN   
1      31266.0     harley  palmer  HR250EJ     False              True  33.0   
2  534142828.0        NaN     NaN      NaN     False              True   NaN   
3       1035.0        NaN     NaN      NaN     False              True   NaN   
4       6985.0        NaN     NaN      NaN     False              True   NaN   

   in_crm_data  
0        False  
1         True  
2        False  
3        False  
4        False

There is now one row per customer from BOTH the CRM data and purchase history

## Customers

In [33]:
customers = pd.read_csv("../data/customer_database.csv")
print(customers.shape)
customers.head()

(23476, 5)


customer_id first_name   surname postcode  age
0         1641       Rhys  Richards  DE456EZ   45
1        24796     Maisie     Young  SW433XX   16
2        14358     Nathan      King   NW49TU   58
3        15306       Jack     Moore  NW908RR   26
4        24971  Alexander   Roberts  SW500HW   85

In [34]:
customers.isnull().sum()

customer_id    0
first_name     0
surname        0
postcode       0
age            0
dtype: int64

In [35]:
for col in ["first_name", "surname"]:
    customers[col] = customers[col].str.lower().str.strip()

customers["postcode"] = customers["postcode"].str.strip()

In [36]:
customers.groupby("customer_id").size().loc[lambda x: x>1]

Series([], dtype: int64)

In [37]:
print("{} rows".format(len(customers)))
unique_customers = customers.drop(columns="customer_id").drop_duplicates()
print("{} unique combinations of customers".format(len(unique_customers)))

23476 rows
19889 unique combinations of customers


Joining this data to our merged customer data

In [38]:
print(len(sales_and_crm_customers))

all_customers = sales_and_crm_customers.merge(customers, on="customer_id", how="left", suffixes=("_sales", "_customers"))
print(len(all_customers))
all_customers.head()

33972
33972


customer_id first_name_sales surname_sales postcode_sales  is_guest  \
0       7466.0              NaN           NaN            NaN     False   
1      31266.0           harley        palmer        HR250EJ     False   
2  534142828.0              NaN           NaN            NaN     False   
3       1035.0              NaN           NaN            NaN     False   
4       6985.0              NaN           NaN            NaN     False   

   in_purchase_data  age_sales  in_crm_data first_name_customers  \
0              True        NaN        False                  eve   
1              True       33.0         True                  NaN   
2              True        NaN        False                  NaN   
3              True        NaN        False                 luca   
4              True        NaN        False                  mia   

  surname_customers postcode_customers  age_customers  
0          richards             HR90PT           45.0  
1               NaN                NaN            NaN  
2               NaN                NaN            NaN  
3            gibson            DE256NH           30.0  
4            rogers            HR662RP           43.0

In [39]:
# now we have duplicate customer detail columns, so merge them into the old ones
merged_customers_filter = (
    (all_customers["customer_id"].notnull()) # only for actual customers
    # only if they have at least first or surname filled in
    & ((all_customers["first_name_customers"].notnull())
       | (all_customers["surname_customers"].notnull()))
)

all_customers.loc[merged_customers_filter, "in_customer_data"] = True
all_customers.loc[~merged_customers_filter, "in_customer_data"] = False

all_customers["in_customer_data"].value_counts()

True     22053
False    11919
Name: in_customer_data, dtype: int64

In [40]:
all_customers.isnull().sum()

customer_id              8300
first_name_sales        17847
surname_sales           17847
postcode_sales          17847
is_guest                    0
in_purchase_data            0
age_sales               26147
in_crm_data                 0
first_name_customers    11919
surname_customers       11919
postcode_customers      11919
age_customers           11919
in_customer_data            0
dtype: int64

If customers are newly added, fill in customer info:

In [41]:
update_filter = (
    (all_customers["in_customer_data"])
    & (all_customers["first_name_sales"].isnull())
    & (all_customers["surname_sales"].isnull())
)

print(len(all_customers))

all_customers.loc[update_filter, ["first_name_sales", "surname_sales", "postcode_sales", "age_sales"]] = (
    all_customers.loc[update_filter, ["first_name_customers", "surname_customers", "postcode_customers", "age_customers"]].values
)

all_customers = (
    all_customers
    .drop(columns=["first_name_customers", "surname_customers", "age_customers", "postcode_customers"])
    .rename(columns={
        "first_name_sales": "first_name",
        "surname_sales": "surname",
        "age_sales": "age",
        "postcode_sales": "postcode"
    })
)

print(len(all_customers))
all_customers.isnull().sum()

33972
33972


customer_id         8300
first_name          1248
surname             1248
postcode            1248
is_guest               0
in_purchase_data       0
age                 9548
in_crm_data            0
in_customer_data       0
dtype: int64

Add customers from customer DB but not in main data

In [42]:
customer_ids_to_add = list(set(customers["customer_id"].unique()) - set(all_customers["customer_id"].unique()))
print(len(customer_ids_to_add))

1423


In [43]:
print(len(all_customers))

all_customers = (
    pd.concat([all_customers, customers[customers["customer_id"].isin(customer_ids_to_add)]],
              axis=0,
              ignore_index=True)
)

print(len(all_customers))

33972
35395


In [44]:
all_customers.isnull().sum()

customer_id         8300
first_name          1248
surname             1248
postcode            1248
is_guest            1423
in_purchase_data    1423
age                 9548
in_crm_data         1423
in_customer_data    1423
dtype: int64

At this point if we have missing data for whether someone was a guest, it means they weren't (since we can only mark guests using the purchase data and we've already done that)

In [45]:
all_customers["is_guest"] = all_customers["is_guest"].fillna(False)
all_customers["in_purchase_data"] = all_customers["in_purchase_data"].fillna(False)
all_customers["in_crm_data"] = all_customers["in_crm_data"].fillna(False)
all_customers["in_customer_data"] = all_customers["in_customer_data"].fillna(True)

all_customers.isnull().sum()

customer_id         8300
first_name          1248
surname             1248
postcode            1248
is_guest               0
in_purchase_data       0
age                 9548
in_crm_data            0
in_customer_data       0
dtype: int64

In [46]:
all_customers.head()

customer_id first_name   surname postcode  is_guest  in_purchase_data  \
0       7466.0        eve  richards   HR90PT     False              True   
1      31266.0     harley    palmer  HR250EJ     False              True   
2  534142828.0        NaN       NaN      NaN     False              True   
3       1035.0       luca    gibson  DE256NH     False              True   
4       6985.0        mia    rogers  HR662RP     False              True   

    age  in_crm_data  in_customer_data  
0  45.0        False              True  
1  33.0         True             False  
2   NaN        False             False  
3  30.0        False              True  
4  43.0        False              True

In [47]:
len(all_customers)

35395

## Deduplicate

Let's first remove exact duplicates.

In [48]:
print(len(all_customers))
all_customers = all_customers.drop_duplicates()
print(len(all_customers))

35395
35395


Let's count up all the following cases:

- those purchased with a customer ID (whether or not that ID is in another database)
- those purchased as guests
- those in the CRM database who haven't made a purchase
- those in the customer database who haven't made a purchase


In [49]:
identified_customers = (
    all_customers[(all_customers["customer_id"].notnull())
                  & (all_customers["in_purchase_data"])
                  & ((all_customers["in_crm_data"])
                     | (all_customers["in_customer_data"]))]
)

guests = all_customers[all_customers["is_guest"]]

customer_ids_not_found = (
    all_customers[(all_customers["customer_id"].notnull())
                  & (all_customers["first_name"].isnull())
                  & (all_customers["surname"].isnull())]
)

customer_data_only = (
    all_customers[((all_customers["in_crm_data"])
                   | (all_customers["in_customer_data"])
                  )
                  & (all_customers["in_purchase_data"] == False)]
)

print(len(all_customers), len(identified_customers))
print(len(guests), len(customer_ids_not_found), len(customer_data_only))

35395 23713
8300 1248 2134


In [50]:
print(f"Size of customer database (not yet deduplicated): {len(all_customers)}")
print(f"Identified customers (with an ID, present in a lookup DB): {len(identified_customers)}")
print(f"Guest checkouts: {len(guests)}")
print(f"Unidentified customer IDs in purchases: {len(customer_ids_not_found)}")
print(f"Customers with no purchases present in CRM or customer data: {len(customer_data_only)}")

Size of customer database (not yet deduplicated): 35395
Identified customers (with an ID, present in a lookup DB): 23713
Guest checkouts: 8300
Unidentified customer IDs in purchases: 1248
Customers with no purchases present in CRM or customer data: 2134


In [51]:
assert len(all_customers) == len(identified_customers) + len(guests) + len(customer_ids_not_found) + len(customer_data_only)


We have around 35000 customers and we need to deal with 2 types of deduplication:
- different customer IDs encoding exactly the same information (exact duplicates)
- different customer IDsencoding almost exactly the same information (fuzzy duplicates)

How do we identify a duplicate?
- different customer ID (or NO customer ID in the case of guests) but the same details
- first name
- surname
- postcode
- age (but this isn't provided for guest accounts!)

We want to have a "main" record where we also record all the IDs that we think refer to the same account.

Guests don't have IDs, let's give them some.

In [52]:
all_guests = all_customers[all_customers["is_guest"]].copy()

# go from -1 to -N as needed
new_ids = np.arange(-1, -(len(all_guests) + 1), -1)

all_customers.loc[all_customers["is_guest"], "customer_id"] = new_ids
all_customers.head()

customer_id first_name   surname postcode  is_guest  in_purchase_data  \
0       7466.0        eve  richards   HR90PT     False              True   
1      31266.0     harley    palmer  HR250EJ     False              True   
2  534142828.0        NaN       NaN      NaN     False              True   
3       1035.0       luca    gibson  DE256NH     False              True   
4       6985.0        mia    rogers  HR662RP     False              True   

    age  in_crm_data  in_customer_data  
0  45.0        False              True  
1  33.0         True             False  
2   NaN        False             False  
3  30.0        False              True  
4  43.0        False              True

In [53]:
all_customers[all_customers["is_guest"]].head()

customer_id first_name    surname postcode  is_guest  in_purchase_data  \
24961         -1.0    michael      mason  RG497ZQ      True              True   
24962         -2.0       cole  wilkinson   SW75TQ      True              True   
24963         -3.0   mohammed   richards  RG150RE      True              True   
24964         -4.0       kian      mills  SW332TF      True              True   
24965         -5.0       ruby       owen  PO377YS      True              True   

       age  in_crm_data  in_customer_data  
24961  NaN        False             False  
24962  NaN        False             False  
24963  NaN        False             False  
24964  NaN        False             False  
24965  NaN        False             False

In [54]:
all_customers["customer_id"] = all_customers["customer_id"].astype(int)

Now we try to deduplicate everything including guest accounts

In [55]:
columns_to_consider = ["first_name", "surname", "postcode"]

duplicates = all_customers[all_customers.duplicated(subset=columns_to_consider, keep=False)]

# Create a dictionary mapping duplicates to customer IDs
duplicate_dict = duplicates.groupby(columns_to_consider)['customer_id'].apply(list).to_dict()

# Add a new column for the duplicated customer IDs
all_customers['other_customer_ids'] = all_customers.apply(lambda x: duplicate_dict.get((x['first_name'], x['surname'], x['postcode'])), axis=1)

all_customers

customer_id first_name   surname postcode  is_guest  in_purchase_data  \
0             7466        eve  richards   HR90PT     False              True   
1            31266     harley    palmer  HR250EJ     False              True   
2        534142828        NaN       NaN      NaN     False              True   
3             1035       luca    gibson  DE256NH     False              True   
4             6985        mia    rogers  HR662RP     False              True   
...            ...        ...       ...      ...       ...               ...   
35390        14295       erin    morgan  NW481EN     False             False   
35391        28025      aaron    harris  SO265RP     False             False   
35392         4220      grace  mitchell  EX709AR     False             False   
35393        13086     oliver      hall  NW277BU     False             False   
35394        29494      reece   chafhan  SO433HB     False             False   

        age  in_crm_data  in_customer_data other_customer_ids  
0      45.0        False              True               None  
1      33.0         True             False      [31266, 5411]  
2       NaN        False             False               None  
3      30.0        False              True               None  
4      43.0        False              True               None  
...     ...          ...               ...                ...  
35390  63.0        False              True     [-8152, 14295]  
35391  66.0        False              True               None  
35392  20.0        False              True               None  
35393  58.0        False              True               None  
35394  41.0        False              True               None  

[35395 rows x 10 columns]

There are records with multiple linked accounts:

In [56]:
all_customers[all_customers["other_customer_ids"].notnull()].head()


customer_id first_name     surname postcode  is_guest  in_purchase_data  \
1         31266     harley      palmer  HR250EJ     False              True   
5         26434     bailey  richardson  SW988AF     False              True   
6         28961       skye     johnson    M80NA     False              True   
10        12586        max       moore   M902XX     False              True   
12        22825     alicia        wood  SO879UN     False              True   

     age  in_crm_data  in_customer_data     other_customer_ids  
1   33.0         True             False          [31266, 5411]  
5   31.0         True              True         [26434, 27761]  
6   54.0        False              True         [28961, 12140]  
10  20.0         True             False  [12586, -6174, 32914]  
12  24.0        False              True         [22825, 28495]

What we want now is for each duplicate combination, mark one record as the "main" so that when we count the "main" records we get a sense of how many unique values there are.

First we need to remove each record from its own duplicate list

In [57]:
def remove_own_record(row):
    ids = list(row["other_customer_ids"])
    ids.remove(row["customer_id"])
    return ids

all_customers.loc[all_customers["other_customer_ids"].notnull(), "duplicate_customer_ids"] = (
    all_customers[all_customers["other_customer_ids"].notnull()].apply(remove_own_record, axis=1)
)

all_customers[all_customers["other_customer_ids"].notnull()].head()

customer_id first_name     surname postcode  is_guest  in_purchase_data  \
1         31266     harley      palmer  HR250EJ     False              True   
5         26434     bailey  richardson  SW988AF     False              True   
6         28961       skye     johnson    M80NA     False              True   
10        12586        max       moore   M902XX     False              True   
12        22825     alicia        wood  SO879UN     False              True   

     age  in_crm_data  in_customer_data     other_customer_ids  \
1   33.0         True             False          [31266, 5411]   
5   31.0         True              True         [26434, 27761]   
6   54.0        False              True         [28961, 12140]   
10  20.0         True             False  [12586, -6174, 32914]   
12  24.0        False              True         [22825, 28495]   

   duplicate_customer_ids  
1                  [5411]  
5                 [27761]  
6                 [12140]  
10         [-6174, 32914]  
12                [28495]

In [58]:
all_customers[all_customers["customer_id"].isin([31266, 5411])]

customer_id first_name surname postcode  is_guest  in_purchase_data  \
1           31266     harley  palmer  HR250EJ     False              True   
7208         5411     harley  palmer  HR250EJ     False              True   

       age  in_crm_data  in_customer_data other_customer_ids  \
1     33.0         True             False      [31266, 5411]   
7208  33.0         True             False      [31266, 5411]   

     duplicate_customer_ids  
1                    [5411]  
7208                [31266]

In [59]:
# first, add a rank per firstname-surname-postcode combination
all_customers["rank"] = all_customers.groupby(columns_to_consider).cumcount()+1

all_customers[all_customers["customer_id"].isin([31266, 5411])]

customer_id first_name surname postcode  is_guest  in_purchase_data  \
1           31266     harley  palmer  HR250EJ     False              True   
7208         5411     harley  palmer  HR250EJ     False              True   

       age  in_crm_data  in_customer_data other_customer_ids  \
1     33.0         True             False      [31266, 5411]   
7208  33.0         True             False      [31266, 5411]   

     duplicate_customer_ids  rank  
1                    [5411]     1  
7208                [31266]     2

In [60]:
all_customers.loc[all_customers["rank"] == 1, "is_main"] = True
all_customers["is_main"] = all_customers["is_main"].fillna(False)

all_customers = all_customers.drop(columns="rank")

print(f"Total customers in DB: {len(all_customers)}")
print(f"Of which {len(all_customers[all_customers['is_main']])} are unique/main records")

Total customers in DB: 35395
Of which 27395 are unique/main records


## Record Linkage

In [62]:
!pip install recordlinkage

import recordlinkage

# index the dataframes
indexer = recordlinkage.Index()
indexer.block('postcode')
candidate_links = indexer.index(all_customers.set_index("customer_id"))

# set up the comparison rules
compare = recordlinkage.Compare()
compare.string('first_name', 'first_name', method='damerau_levenshtein', threshold=0.85, label="first_name")
compare.string('surname', 'surname', method='damerau_levenshtein', threshold=0.85, label="surname")
compare.exact('postcode', 'postcode', label="postcode")

# create the comparison vectors
compare_vectors = compare.compute(candidate_links, all_customers.set_index("customer_id"))

This is now every combination of customer IDs and how many criteria matched

In [63]:
compare_vectors

first_name  surname  postcode
customer_id_1 customer_id_2                               
7523           7466                 0.0      0.0         1
7492           7466                 0.0      0.0         1
               7523                 0.0      0.0         1
7518           7466                 0.0      0.0         1
               7523                 0.0      0.0         1
...                                 ...      ...       ...
17659         -7621                 1.0      1.0         1
              -7924                 0.0      0.0         1
               17654                0.0      0.0         1
               17680                0.0      0.0         1
               27928                0.0      0.0         1

[1020864 rows x 3 columns]

Extract just the perfect matches, i.e. 3/3 criteria passed

In [64]:
matches = compare_vectors[compare_vectors.sum(axis=1) == 3]
matches

first_name  surname  postcode
customer_id_1 customer_id_2                               
 7469          28769                1.0      1.0         1
 28583         7468                 1.0      1.0         1
 32670         7485                 1.0      1.0         1
 31690         7476                 1.0      1.0         1
 32600         7516                 1.0      1.0         1
...                                 ...      ...       ...
 17677         32903                1.0      1.0         1
-489           30716                1.0      1.0         1
-2407          17682                1.0      1.0         1
 17680        -1098                 1.0      1.0         1
 17659        -7621                 1.0      1.0         1

[7148 rows x 3 columns]

Now merge this back to the customer data so we have all the customer records alongside the duplicate pairs

In [65]:
match_df = pd.DataFrame(
    data=matches.index.tolist(),
    columns=["customer_id_1", "customer_id_2"]
)

matched = all_customers.merge(match_df, left_on="customer_id", right_on="customer_id_1", how="left", suffixes=("_customers", "_matches"))
matched = matched.merge(match_df, left_on="customer_id", right_on="customer_id_2", how="left", suffixes=("_customers", "_matches"))

matched_columns_to_show = ["customer_id", "first_name", "surname", "postcode", "other_customer_ids", "duplicate_customer_ids", "is_main", "customer_id_1_customers", "customer_id_2_customers", "customer_id_1_matches", "customer_id_2_matches"]
matched[matched_columns_to_show]

customer_id first_name   surname postcode other_customer_ids  \
0             7466        eve  richards   HR90PT               None   
1            31266     harley    palmer  HR250EJ      [31266, 5411]   
2        534142828        NaN       NaN      NaN               None   
3             1035       luca    gibson  DE256NH               None   
4             6985        mia    rogers  HR662RP               None   
...            ...        ...       ...      ...                ...   
35804        14295       erin    morgan  NW481EN     [-8152, 14295]   
35805        28025      aaron    harris  SO265RP               None   
35806         4220      grace  mitchell  EX709AR               None   
35807        13086     oliver      hall  NW277BU               None   
35808        29494      reece   chafhan  SO433HB               None   

      duplicate_customer_ids  is_main  customer_id_1_customers  \
0                        NaN     True                      NaN   
1                     [5411]     True                      NaN   
2                        NaN     True                      NaN   
3                        NaN     True                      NaN   
4                        NaN     True                      NaN   
...                      ...      ...                      ...   
35804                [-8152]    False                  14295.0   
35805                    NaN     True                      NaN   
35806                    NaN     True                   4220.0   
35807                    NaN     True                      NaN   
35808                    NaN     True                      NaN   

       customer_id_2_customers  customer_id_1_matches  customer_id_2_matches  
0                          NaN                    NaN                    NaN  
1                          NaN                 5411.0                31266.0  
2                          NaN                    NaN                    NaN  
3                          NaN                    NaN                    NaN  
4                          NaN                    NaN                    NaN  
...                        ...                    ...                    ...  
35804                  -8152.0                    NaN                    NaN  
35805                      NaN                    NaN                    NaN  
35806                  30722.0                    NaN                    NaN  
35807                      NaN                    NaN                    NaN  
35808                      NaN                    NaN                    NaN  

[35809 rows x 11 columns]

In [66]:
matched.loc[matched["customer_id"] == 30730, matched_columns_to_show]

customer_id first_name    surname postcode  \
4829        30730     harvey  robertson  NW436BL   
4830        30730     harvey  robertson  NW436BL   
4831        30730     harvey  robertson  NW436BL   

                other_customer_ids duplicate_customer_ids  is_main  \
4829  [30730, 28816, 13817, 13814]  [28816, 13817, 13814]     True   
4830  [30730, 28816, 13817, 13814]  [28816, 13817, 13814]     True   
4831  [30730, 28816, 13817, 13814]  [28816, 13817, 13814]     True   

      customer_id_1_customers  customer_id_2_customers  customer_id_1_matches  \
4829                      NaN                      NaN                28816.0   
4830                      NaN                      NaN                13817.0   
4831                      NaN                      NaN                13814.0   

      customer_id_2_matches  
4829                30730.0  
4830                30730.0  
4831                30730.0

Reduce multiple records into a single list (like duplicate_customer_ids)

In [67]:
def merge_duplicates(group):
    duplicate_list = []
    if np.isnan(group["customer_id_1_matches"].values[0]) == False:
        duplicate_list.extend(group["customer_id_1_matches"].tolist())
    if np.isnan(group["customer_id_2_customers"].values[0]) == False:
        duplicate_list.extend(group["customer_id_2_customers"].tolist())
    if len(duplicate_list) > 0:
        return sorted(list(set([int(x) for x in duplicate_list])))
    return np.nan

linkages = (
    matched
    .groupby("customer_id")
    .apply(merge_duplicates)
    .reset_index(name="linked_duplicates")
)

linkages.head()

customer_id linked_duplicates
0        -8300               NaN
1        -8299            [5652]
2        -8298               NaN
3        -8297               NaN
4        -8296               NaN

Merge this back to the main dataset to see the difference

In [68]:
all_customers = all_customers.merge(linkages, on="customer_id", how="left")
all_customers.head()

customer_id first_name   surname postcode  is_guest  in_purchase_data  \
0         7466        eve  richards   HR90PT     False              True   
1        31266     harley    palmer  HR250EJ     False              True   
2    534142828        NaN       NaN      NaN     False              True   
3         1035       luca    gibson  DE256NH     False              True   
4         6985        mia    rogers  HR662RP     False              True   

    age  in_crm_data  in_customer_data other_customer_ids  \
0  45.0        False              True               None   
1  33.0         True             False      [31266, 5411]   
2   NaN        False             False               None   
3  30.0        False              True               None   
4  43.0        False              True               None   

  duplicate_customer_ids  is_main linked_duplicates  
0                    NaN     True               NaN  
1                 [5411]     True            [5411]  
2                    NaN     True               NaN  
3                    NaN     True               NaN  
4                    NaN     True               NaN

Sort original duplicate customer IDs too so we can compare results

In [69]:
all_customers["duplicate_customer_ids"] = all_customers.loc[all_customers["duplicate_customer_ids"].notnull(), "duplicate_customer_ids"].apply(lambda x: sorted(x))


Where do the methods disagree?

In [70]:
(
    all_customers[(all_customers["duplicate_customer_ids"].notnull())
               & (all_customers["linked_duplicates"].notnull())
               & (all_customers["duplicate_customer_ids"] != all_customers["linked_duplicates"])]
)

customer_id first_name   surname postcode  is_guest  in_purchase_data  \
2197         10383   scarlett   jackson   M284ZB     False              True   
7203         19549     summer  anderson  RG546PY     False              True   
9445         16969    georgia     scott  PO466DY     False              True   
13277         5390       josh   simpson  HR235FS     False              True   
14692        32354       josh   simpson  HR235FS     False              True   
16230        19551     summer  anderson  RG546PY     False              True   
21022        16966    georgia     scott  PO466DY     False              True   
24809        10386   scarlett   jackson   M284ZB     False              True   
28568        -3608   isabella  harrison  NW908RT      True              True   
31139        -6179       lily   chapman  LS238QF      True              True   
33507        15377   isabella  harrison  NW908RT     False             False   
33967         7936       lily   chapman  LS238QF     False             False   
34662         7934       lily   chapman  LS238QF     False             False   

        age  in_crm_data  in_customer_data   other_customer_ids  \
2197   35.0         True              True       [10383, 10386]   
7203   47.0         True             False       [19549, 19551]   
9445    9.0        False              True       [16969, 16966]   
13277  81.0        False              True        [5390, 32354]   
14692  81.0        False              True        [5390, 32354]   
16230  38.0        False              True       [19549, 19551]   
21022  44.0        False              True       [16969, 16966]   
24809  12.0        False              True       [10383, 10386]   
28568   NaN        False             False       [-3608, 15377]   
31139   NaN        False             False  [-6179, 7936, 7934]   
33507   5.0         True              True       [-3608, 15377]   
33967  28.0         True              True  [-6179, 7936, 7934]   
34662  56.0        False              True  [-6179, 7936, 7934]   

      duplicate_customer_ids  is_main      linked_duplicates  
2197                 [10386]     True  [10386, 27536, 28786]  
7203                 [19551]     True         [19551, 28832]  
9445                 [16966]     True         [16966, 32299]  
13277                [32354]     True         [32354, 33132]  
14692                 [5390]    False          [5390, 33132]  
16230                [19549]    False         [19549, 28832]  
21022                [16969]    False         [16969, 32299]  
24809                [10383]    False  [10383, 27536, 28786]  
28568                [15377]     True         [15376, 15377]  
31139           [7934, 7936]     True    [7934, 7936, 32787]  
33507                [-3608]    False         [-3608, 15376]  
33967          [-6179, 7934]    False   [-6179, 7934, 32787]  
34662          [-6179, 7936]    False   [-6179, 7936, 32787]

In [71]:
all_customers[all_customers["customer_id"].isin([10383, 10386, 27536, 28786])]

customer_id first_name  surname postcode  is_guest  in_purchase_data  \
2197         10383   scarlett  jackson   M284ZB     False              True   
10596        27536   scariett  jackson   M284ZB     False              True   
19538        28786   sgarlett  jagkson   M284ZB     False              True   
24809        10386   scarlett  jackson   M284ZB     False              True   

        age  in_crm_data  in_customer_data other_customer_ids  \
2197   35.0         True              True     [10383, 10386]   
10596  12.0        False              True               None   
19538  35.0        False              True               None   
24809  12.0        False              True     [10383, 10386]   

      duplicate_customer_ids  is_main      linked_duplicates  
2197                 [10386]     True  [10386, 27536, 28786]  
10596                    NaN     True         [10383, 10386]  
19538                    NaN     True         [10383, 10386]  
24809                [10383]    False  [10383, 27536, 28786]

There we have it. Our modeled data.